File: BE/app/api/v1/endpoints/courses.py
"""Course API endpoints — student-facing course content APIs."""

from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.postgres import get_db
from app.services.course_service import CourseService
from app.schemas.course import (
    CourseListResponse, CourseDetail,
    LessonOut, EnrollRequest, EnrollmentOut,
    ProgressUpdate, LessonProgressOut,
    QuizOut, QuizSubmission, QuizResultOut,
    MaterialOut, FlashcardDeckOut, CategoryOut,
)

router = APIRouter(prefix="/courses", tags=["Courses"])


def get_service(db: AsyncSession = Depends(get_db)) -> CourseService:
    return CourseService(db)


# ── Course Discovery ──

@router.get("", response_model=CourseListResponse)
async def list_courses(
    page: int = Query(1, ge=1),
    page_size: int = Query(12, ge=1, le=50),
    category: Optional[str] = Query(None, description="Filter by category slug"),
    difficulty: Optional[str] = Query(None, description="beginner|intermediate|advanced"),
    pricing: Optional[str] = Query(None, description="free|one_time|subscription"),
    search: Optional[str] = Query(None, description="Search in title/description"),
    service: CourseService = Depends(get_service),
):
    """List published courses with filters and pagination."""
    return await service.list_courses(
        page=page, page_size=page_size,
        category=category, difficulty=difficulty,
        pricing=pricing, search=search,
    )


@router.get("/categories", response_model=list[CategoryOut])
async def list_categories(service: CourseService = Depends(get_service)):
    """List all active course categories."""
    return await service.get_categories()


@router.get("/{slug}", response_model=CourseDetail)
async def get_course(slug: str, service: CourseService = Depends(get_service)):
    """Get full course detail with curriculum (modules + lessons)."""
    course = await service.get_course_detail(slug)
    if not course:
        raise HTTPException(status_code=404, detail="Course not found")
    return course


# ── Lesson Content ──

@router.get("/lessons/{lesson_id}", response_model=LessonOut)
async def get_lesson(lesson_id: int, service: CourseService = Depends(get_service)):
    """Get lesson content (video, text, PDF URL, etc.)."""
    lesson = await service.get_lesson(lesson_id)
    if not lesson:
        raise HTTPException(status_code=404, detail="Lesson not found")
    return lesson


# ── Quiz ──

@router.get("/quizzes/{quiz_id}", response_model=QuizOut)
async def get_quiz(quiz_id: int, service: CourseService = Depends(get_service)):
    """Get quiz with questions (correct answers stripped)."""
    quiz = await service.get_quiz(quiz_id)
    if not quiz:
        raise HTTPException(status_code=404, detail="Quiz not found")
    return quiz


@router.post("/quizzes/{quiz_id}/submit", response_model=QuizResultOut)
async def submit_quiz(
    quiz_id: int,
    submission: QuizSubmission,
    enrollment_id: int = Query(..., description="Enrollment ID"),
    service: CourseService = Depends(get_service),
):
    """Submit quiz answers and get graded result."""
    try:
        return await service.submit_quiz(quiz_id, enrollment_id, submission)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ── Materials ──

@router.get("/{course_id}/materials", response_model=list[MaterialOut])
async def get_materials(course_id: int, service: CourseService = Depends(get_service)):
    """Get downloadable materials for a course."""
    return await service.get_materials(course_id)


# ── Flashcards ──

@router.get("/flashcards/{deck_id}", response_model=FlashcardDeckOut)
async def get_flashcard_deck(
    deck_id: int, service: CourseService = Depends(get_service)
):
    """Get flashcard deck with all cards."""
    deck = await service.get_flashcard_deck(deck_id)
    if not deck:
        raise HTTPException(status_code=404, detail="Flashcard deck not found")
    return deck


# ── Student Endpoints (Enrollment & Progress) ──

student_router = APIRouter(prefix="/students", tags=["Student Learning"])


@student_router.post("/enroll/{course_id}", response_model=EnrollmentOut)
async def enroll_in_course(
    course_id: int,
    body: EnrollRequest,
    service: CourseService = Depends(get_service),
):
    """Enroll a student in a course."""
    try:
        return await service.enroll_student(body.student_id, course_id)
    except ValueError as e:
        raise HTTPException(status_code=409, detail=str(e))


@student_router.get("/enrollments", response_model=list[EnrollmentOut])
async def get_enrollments(
    student_id: int = Query(..., description="Student ID"),
    service: CourseService = Depends(get_service),
):
    """Get all enrollments for a student."""
    return await service.get_enrollments(student_id)


@student_router.post(
    "/enrollments/{course_id}/progress",
    response_model=LessonProgressOut,
)
async def update_progress(
    course_id: int,
    body: ProgressUpdate,
    student_id: int = Query(..., description="Student ID"),
    service: CourseService = Depends(get_service),
):
    """Update lesson progress (video position, completion, time spent)."""
    try:
        return await service.update_progress(student_id, course_id, body)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@student_router.get(
    "/enrollments/{course_id}/progress",
    response_model=list[LessonProgressOut],
)
async def get_progress(
    course_id: int,
    student_id: int = Query(..., description="Student ID"),
    service: CourseService = Depends(get_service),
):
    """Get all lesson progress for a student's enrollment."""
    return await service.get_course_progress(student_id, course_id)

File: UI/lib/api/courses.ts
import { fetchApi, fetchApiWithAuth, API_BASE_URL } from './client';
import { CourseListItem, CourseListResponse, CourseDetail, EnrollmentOut, ProgressUpdate, LessonProgressOut, CourseReview, ReviewStats, MaterialOut, AdminCourse, QuizResultOut } from './types';
import * as T from './types';

function capitalize(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export async function getCourses(params?: {
  page?: number;
  page_size?: number;
  category?: string;
  difficulty?: string;
  pricing?: string;
  search?: string;
}): Promise<CourseListResponse> {
  const query = new URLSearchParams();
  if (params?.page) query.set("page", String(params.page));
  if (params?.page_size) query.set("page_size", String(params.page_size));
  if (params?.category) query.set("category", params.category);
  if (params?.difficulty) query.set("difficulty", params.difficulty);
  if (params?.pricing) query.set("pricing", params.pricing);
  if (params?.search) query.set("search", params.search);

  const qs = query.toString();
  return fetchApi<CourseListResponse>(`/courses${qs ? `?${qs}` : ""}`);
}

export async function getCategories(): Promise<T.Category[]> {
  return fetchApi<T.Category[]>("/courses/categories");
}

export async function getCourseDetail(slug: string): Promise<CourseDetail> {
  return fetchApi<CourseDetail>(`/courses/${slug}`);
}

export async function getLesson(lessonId: number): Promise<T.LessonFull> {
  return fetchApi<T.LessonFull>(`/courses/lessons/${lessonId}`);
}

export async function getQuiz(quizId: number): Promise<T.QuizOut> {
  return fetchApi<T.QuizOut>(`/courses/quizzes/${quizId}`);
}

export async function getCourseMaterials(courseId: number): Promise<MaterialOut[]> {
  return fetchApi<MaterialOut[]>(`/courses/${courseId}/materials`);
}

export async function getAllMaterials(): Promise<(MaterialOut & { courseName: string })[]> {
  return fetchApiWithAuth<(MaterialOut & { courseName: string })[]>("/students/materials/all");
}

export async function getFlashcardDeck(deckId: number): Promise<T.FlashcardDeckOut> {
  return fetchApi<T.FlashcardDeckOut>(`/courses/flashcards/${deckId}`);
}

export async function getEnrollments(studentId: number): Promise<EnrollmentOut[]> {
  return fetchApi<EnrollmentOut[]>(`/students/enrollments?student_id=${studentId}`);
}

export async function enrollInCourse(
  courseId: number,
  studentId: number
): Promise<EnrollmentOut> {
  return fetchApiWithAuth<EnrollmentOut>(`/students/enroll/${courseId}`, {
    method: "POST",
    body: JSON.stringify({ student_id: studentId }),
  });
}

export async function getCourseProgress(studentId: number, courseId: number): Promise<LessonProgressOut[]> {
  return fetchApiWithAuth<LessonProgressOut[]>(`/students/enrollments/${courseId}/progress?student_id=${studentId}`);
}

export async function updateLessonProgress(
  studentId: number, 
  courseId: number, 
  progress: ProgressUpdate
): Promise<LessonProgressOut> {
  return fetchApiWithAuth<LessonProgressOut>(`/students/enrollments/${courseId}/progress?student_id=${studentId}`, {
    method: "POST",
    body: JSON.stringify(progress),
  });
}

export async function submitQuiz(
  quizId: number,
  enrollmentId: number,
  answers: Record<string, number | string>
): Promise<QuizResultOut> {
  return fetchApiWithAuth<QuizResultOut>(
    `/courses/quizzes/${quizId}/submit?enrollment_id=${enrollmentId}`,
    {
      method: "POST",
      body: JSON.stringify({ answers }),
    }
  );
}

export function mapCourseToUI(course: CourseListItem) {
  const price = parseFloat(course.price);
  const isFree = course.pricing_model === "free" || price === 0;
  const instructorName = course.instructor
    ? `${course.instructor.first_name} ${course.instructor.last_name}`
    : "Unknown";

  return {
    id: course.course_id,
    title: course.title,
    slug: course.slug,
    category: course.category?.name || "General",
    level: capitalize(course.difficulty_level),
    price: isFree ? 0 : price,
    rating: parseFloat(course.average_rating) || 0,
    students: course.total_enrollments,
    instructor: instructorName,
    image: course.thumbnail_url || "",
    duration: course.duration_hours ? `${course.duration_hours} hours` : "Self-paced",
    modules: course.total_modules,
    progress: 0, // Would come from enrollment/progress API
    isFree,
    skills: course.skills,
    shortDescription: course.short_description,
  };
}

export function mapMaterialToUI(material: MaterialOut & { courseName: string }) {
  const sizeKB = material.file_size_bytes
    ? material.file_size_bytes > 1024 * 1024
      ? `${(material.file_size_bytes / (1024 * 1024)).toFixed(1)} MB`
      : `${(material.file_size_bytes / 1024).toFixed(0)} KB`
    : "—";

  return {
    id: material.material_id,
    title: material.title,
    type: material.file_type,
    course: material.courseName,
    size: sizeKB,
    downloads: material.download_count,
    fileUrl: material.file_url,
    description: material.description,
  };
}

export async function getAdminCourses(params: {
  search?: string;
  status?: string;
  sort_by?: string;
  sort_order?: string;
  limit?: number;
  offset?: number;
}): Promise<{ courses: AdminCourse[]; total: number; summary: { published_count: number; draft_count: number; total_enrollments: number } }> {
  const query = new URLSearchParams();
  if (params.search) query.set("search", params.search);
  if (params.status) query.set("status", params.status);
  if (params.sort_by) query.set("sort_by", params.sort_by);
  if (params.sort_order) query.set("sort_order", params.sort_order);
  if (params.limit) query.set("limit", String(params.limit));
  if (params.offset) query.set("offset", String(params.offset));
  return fetchApiWithAuth(`/admin/courses?${query.toString()}`);
}

export async function toggleCoursePublish(courseId: number, publish: boolean): Promise<{ success: boolean }> {
  return fetchApiWithAuth(`/admin/courses/${courseId}/publish?publish=${publish}`, { method: "PUT" });
}

export async function deleteAdminCourse(courseId: number): Promise<{ success: boolean }> {
  return fetchApiWithAuth(`/admin/courses/${courseId}`, { method: "DELETE" });
}

export async function getCourseReviews(courseId: number, params?: {
  sort_by?: string;
  limit?: number;
  offset?: number;
}): Promise<{ reviews: CourseReview[]; stats: ReviewStats }> {
  const query = new URLSearchParams();
  if (params?.sort_by) query.set("sort_by", params.sort_by);
  if (params?.limit) query.set("limit", String(params.limit));
  if (params?.offset) query.set("offset", String(params.offset));
  const res = await fetch(`${API_BASE_URL}/courses/${courseId}/reviews?${query.toString()}`);
  if (!res.ok) throw new Error("Failed to fetch reviews");
  return res.json();
}

export async function getMyReview(courseId: number): Promise<{ review: { review_id: number; rating: number; review_text: string; created_at: string } | null }> {
  return fetchApiWithAuth(`/courses/${courseId}/reviews/me`);
}

export async function submitReview(courseId: number, rating: number, reviewText: string): Promise<{ success: boolean; review_id: number }> {
  const query = new URLSearchParams({ rating: String(rating), review_text: reviewText });
  return fetchApiWithAuth(`/courses/${courseId}/reviews?${query.toString()}`, { method: "POST" });
}

export async function updateReview(courseId: number, rating: number, reviewText: string): Promise<{ success: boolean }> {
  const query = new URLSearchParams({ rating: String(rating), review_text: reviewText });
  return fetchApiWithAuth(`/courses/${courseId}/reviews?${query.toString()}`, { method: "PUT" });
}

export async function deleteReview(courseId: number): Promise<{ success: boolean }> {
  return fetchApiWithAuth(`/courses/${courseId}/reviews`, { method: "DELETE" });
}

export function loadRazorpayScript(): Promise<boolean> {
  return new Promise((resolve) => {
    if (typeof window === "undefined") return resolve(false);
    if ((window as any).Razorpay) return resolve(true);

    const script = document.createElement("script");
    script.src = "https://checkout.razorpay.com/v1/checkout.js";
    script.async = true;
    script.onload = () => resolve(true);
    script.onerror = () => resolve(false);
    document.body.appendChild(script);
  });
}
